### ISSUE TITLE
A character outside of a comment is painted over

### ISSUE THREAD

**Issue by DimaKiva** [godotengine/godot#63605]
### Godot version

4.0 (alpha 13)

### System information

Ubuntu 18.04

### Issue description

![Снимок экрана от 2022-07-29 09-56-46](https://user-images.githubusercontent.com/11325733/181673622-052c64c2-b63b-43ce-ae53-420f06c65cce.png)

### Steps to reproduce

`var test = []# test`

**Comment by MewPurPur** [comment]
It's a broader issue of unwanted ligatures screwing with the highlighting.

`]#` becomes considered as one character, similar to `:=` or `&&`. Then the entire ligature steals the comment color.

I don't know of any other instances of ligatures messing up highlighting on valid syntax atm, however, if my #63326 gets merged, `--` and `++` will join the party.

**Comment by akien-mga** [comment]
CC @Paulb23 @bruvzg

**Comment by bruvzg** [comment]
I do not think we can do anything with it, ligatures are either on or off (some fonts include style sets to switch some ligatures individually, but not JetBrainsMono).

**Comment by bruvzg** [comment]
Actually, `CodeEdit`/`TextEdit` can break ligatures by submitting parts of the string as a separate spans to the `TextServer`:

But I'm not sure what's the best way to do it, something like will work (obviously not a proper solution, just a POC):
```diff

diff --git a/scene/gui/text_edit.cpp b/scene/gui/text_edit.cpp
index 630a3316d6..46ffa30737 100644
--- a/scene/gui/text_edit.cpp
+++ b/scene/gui/text_edit.cpp
@@ -190,7 +190,16 @@ void TextEdit::Text::invalidate_cache(int p_line, int p_column, bool p_text_chan
 		}
 	} else {
 		if (p_text_changed) {
-			text.write[p_line].data_buf->add_string(text[p_line].data, font, font_size, language);
+			const String &txt = text[p_line].data;
+			int i = txt.find("]#");
+			int e = 0;
+			while (i >= 0) {
+				text.write[p_line].data_buf->add_string(txt.substr(e, i + 1 - e), font, font_size, language);	
+				e = i + 1;
+				i = txt.find("]#", i + 1);
+			}
+			text.write[p_line].data_buf->add_string(txt.substr(e, -1), font, font_size, language);	
+			//text.write[p_line].data_buf->add_string(text[p_line].data, font, font_size, language);
 		}
 		if (!text[p_line].bidi_override.is_empty()) {
 			TS->shaped_text_set_bidi_override(text.write[p_line].data_buf->get_rid(), text[p_line].bidi_override);
```

<img width="367" alt="Screenshot 2022-07-29 at 13 09 49" src="https://user-images.githubusercontent.com/7645683/181737328-c26997cd-d37b-4987-92dc-cae273386fe9.png">

**Comment by bruvzg** [comment]
This will require code highlighting to be integrated at the text shaping step, instead of the drawing step. And as a free bonus, it allows using different fonts for highlighting (like highlighting with a bold or italic instead of color).

**Comment by Calinou** [comment]
In the meantime, I recommend always adding a space before inline `#` comments (or ideally two, like in Python's code style guidelines).

**Comment by MewPurPur** [comment]
Now that #63326 has been merged, I identified 25 ligatures in the default font that can possibly cause incorrect highlighting. I believe this list is complete.

Changed up the highlighting of numbers a little to make the issue more clear:

![image](https://user-images.githubusercontent.com/85438892/185598673-8db844e0-9535-4fcc-b371-1577baafec3b.png)

Without ligatures, all of these look properly:

![image](https://user-images.githubusercontent.com/85438892/185598780-df89a66b-25a5-4a9b-a7cf-5072e0620ba4.png)

Here's the code from above, just copy it into a script:

```
var A = []# '#' shouldn't highlight something before it
var B = 6 <#-- Same mistake as above
~@export var C   # '@' shouldn't highlight something before it
var D = 99 <$Mario.coins   # '$' shouldn't highlight something before it
var E = &=   # Unary '&' iss highlighted like StringName as it expects a string
var F = ^=   # Same mistake as above, but with '^' and NodePath
var G = true &&&"Lmofa" == &"Lmfao"
var H = 6 <+9
var I = 6 <-9
var J = 6 <--9
var K = 6 >-9
var L = 6 ++9
var M = 6 +++9
var N = 6 --9
var O = 6 ---9
var P = 6 ~~9   # The nonnexistent binary '~' uses symbol color
var Q = 6 ~-9   # Same mistake as above
var R = 6 -~9
var S = 6 <~9
var T = 6 <~~9
var U = 6 |-(-9)
var V = 6 <<-(-9)
var W = 6 >>-(-9)
var X = false ||-6 < 9
var Y = false <!--6 < 9
```

**Comment by bruvzg** [comment]
> This will require code highlighting to be integrated at the text shaping step, instead of the drawing step.

Experimented with this a bit. Highlighter seems to be invalidation all lines from the previous edited one to the end of the document, so moving it to the shape step directly will significantly decrease performance, so probably this will require making highlighter asynchronous to be usable.
